modules/host/negotiate.go

```diff
@@ -15,280 +15,356 @@ import (
 	h.save()
 	h.mu.Unlock(lockID)
 
-	// Send an ack to the renter that all is well.
-	err = encoding.WriteObject(conn, true)
+	return nil
+}
+
+// rpcRevise is an RPC that allows a renter to revise a file contract. It will
+// read new revisions in a loop until the renter sends a termination signal.
+func (h *Host) rpcRevise(conn net.Conn) error {
+	// read ID of contract to be revised
+	var fcid types.FileContractID
+	if err := encoding.ReadObject(conn, &fcid, crypto.HashSize); err != nil {
+		return err
+	}
+	lockID := h.mu.RLock()
+	obligation, exists := h.obligationsByID[fcid]
+	h.mu.RUnlock(lockID)
+	if !exists {
+		return errors.New("no record of that contract")
+	}
+
+	// need to protect against two simultaneous revisions to the same
+	// contract; this can cause inconsistency and data loss, making storage
+	// proofs impossible
+	obligation.mu.Lock()
+	defer obligation.mu.Unlock()
+
+	// open the file (or create it if it does not exist)
+	file, err := os.OpenFile(obligation.Path, os.O_WRONLY|os.O_APPEND, 0660)
 	if err != nil {
-		return
+		return err
+	}
+
+	// rebuild current Merkle tree
+	tree := crypto.NewTree()
+	buf := make([]byte, crypto.SegmentSize)
+	for {
+		_, err := io.ReadFull(file, buf)
+		if err == io.EOF {
+			break
+		} else if err != nil && err != io.ErrUnexpectedEOF {
+			return err
+		}
+		tree.Push(buf)
+	}
+
+	// accept new revisions in a loop
+	for {
+		// read proposed revision
+		var rev types.FileContractRevision
+		if err := encoding.ReadObject(conn, &rev, types.BlockSizeLimit); err != nil {
+			return err
+		}
+		// an empty revision indicates completion
+		if rev.ParentID == (types.FileContractID{}) {
+			break
+		}
+
+		// check revision against original file contract
+		lockID = h.mu.RLock()
+		err := h.considerRevision(rev, obligation)
+		height := h.blockHeight
+		h.mu.RUnlock(lockID)
+		if err != nil {
+			encoding.WriteObject(conn, err.Error())
+			continue // don't terminate loop; subsequent revisions may be okay
+		}
+
+		// indicate acceptance
+		if err := encoding.WriteObject(conn, modules.AcceptResponse); err != nil {
+			return err
+		}
+
+		// read piece
+		piece := make([]byte, rev.NewFileSize-obligation.FileContract.FileSize)
+		_, err = io.ReadFull(conn, piece)
+		if err != nil {
+			return err
+		}
+
+		// verify Merkle root
+		tree.Push(piece)
+		if tree.Root() != rev.NewFileMerkleRoot {
+			return errors.New("revision has bad Merkle root")
+		}
+
+		// read signed transaction
+		var signedTxn types.Transaction
+		if err := encoding.ReadObject(conn, &signedTxn, types.BlockSizeLimit); err != nil {
+			return err
+		}
+
+		// check signature
+		if err := signedTxn.StandaloneValid(height); err != nil {
+			return err
+		}
+
+		// sign and return transaction
+		txnBuilder := h.wallet.RegisterTransaction(signedTxn, nil)
+		for _, sig := range signedTxn.TransactionSignatures {
+			txnBuilder.AddTransactionSignature(sig)
+		}
+		signedTxnSet, err := txnBuilder.Sign(true)
```

Comment:
this won't work.

The txn builder can only sign inputs it funded itself. Since you didn't use FundTransaction, this is basically a no-op.

To sign the transcation you're going to need the secret key you made earlier. The process is a bit convoluted because you sign everything except the signature itself, which means you have to craft the signature items. As a reference, checkout transactionbuilder.Sign.

Here's a somewhat condensed version of the code you'll need:

```
    coveredFields := types.CoveredFields{WholeTransaction: true}

    // TransactionSignatures don't get covered by the 'WholeTransaction' flag,
    // and must be covered manually.
    for i := range txn.TransactionSignatures {
        coveredFields.TransactionSignatures = append(coveredFields.TransactionSignatures, uint64(i))
    }

         // Create the unsigned transaction signature.
        sig := types.TransactionSignature{
            ParentID:       parentID,
            CoveredFields:  coveredFields,
            PublicKeyIndex: 1, // depends on if the host key is 1st or second, you'll have to figure that out
        }
        txn.TransactionSignatures = append(txn.TransactionSignatures, sig) // txn not signed yet, just prepared

        // Get the signature.
        sigIndex := len(txn.TransactionSignatures) - 1
        sigHash := txn.SigHash(sigIndex) // this will figure out what hash is supposed to be signed using coveredFields and such
        encodedSig, err := crypto.SignHash(sigHash,  secretKey[i]) // actual sig
        if err != nil {
            return err
        }
        txn.TransactionSignatures[sigIndex].Signature = encodedSig[:]
    }


```